packages/website/blog/2024-03-25-changes-to-consistent-type-imports-with-decorators.md

```diff
@@ -0,0 +1,101 @@
+---
+authors:
+  - image_url: /img/team/bradzacher.jpg
+    name: Brad Zacher
+    title: typescript-eslint Maintainer
+    url: https://github.com/bradzacher
+description: Changes to consistent-type-imports when used with decorators, experimentalDecorators, and emitDecoratorMetadata
+slug: changes-to-consistent-type-imports-with-decorators
+tags:
+  [
+    consistent-type-imports,
+    experimentalDecorators,
+    emitDecoratorMetadata,
+    typescript-eslint,
+  ]
+title: Changes to consistent-type-imports when used with legacy decorators and decorator metadata
+---
+
+We've made some changes to `consistent-type-imports` to fix some long-standing issues when used alongside `experimentalDecorators: true` and `emitDecoratorMetadata: true`. These changes increase safety and prevent invalid fixes when using decorator metadata.
+
+<!-- truncate -->
+
+## Experimental Decorator Metadata
+
+The `experimentalDecorators` compiler option (referred to as "legacy decorators" from here on) turns on support for an old version of the decorator proposal that was never standardized. Syntactically this proposal isn't too dissimilar to the current [Stage 3 proposal](https://github.com/tc39/proposal-decorators), however they differ in one major way - the use of [metadata reflection](https://rbuckton.github.io/reflect-metadata/) when `emitDecoratorMetadata` is turned on.
+
+When using legacy decorators with decorator metadata and a class is annotated with decorators TypeScript will emit runtime metadata for the class that captures the property types, method parameter types, and method return types. This metadata is incredibly powerful and provides a bridge between the types (which are elided at compile time) and the runtime code. However this runtime code is generated at a cost as it is derived using type information - meaning that the runtime code emitted changes based on the cross-file type information that TypeScript has computed.
```

Comment:
Suggestion, I'd reword a bit to split up sentences and the paragraphs into two discrete points, and be slightly less flattering for the decorators:

```suggestion
When using legacy decorators with decorator metadata and a class is annotated with decorators, TypeScript will emit runtime metadata for the class. That decorator metadata will capture property types, method parameter types, and method return types. Decorator metadata provides a bridge between the types (which are not available at compile time) and the runtime code. 

The downside of generating this runtime code that it is derived using type information: meaning that the runtime code emitted changes based on the cross-file type information that TypeScript has computed. Doing so violates a key [TypeScript design goal](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals) of not changing runtime behavior based on type information.
```